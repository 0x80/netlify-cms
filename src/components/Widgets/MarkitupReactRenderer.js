import React, { PropTypes } from 'react';
import MarkupIt, { Syntax, JSONUtils } from 'markup-it';

const defaultRenderers = {
  'doc': 'article',
  'header_one': 'h1',
  'header_two': 'h2',
  'header_three': 'h3',
  'header_four': 'h4',
  'header_five': 'h5',
  'header_six': 'h6',
  'paragraph': 'p',
  'ordered_list': 'ol',
  'unordered_list': 'ul',
  'list_item': 'li',
  'link': 'a',
  'image': 'img',
  'BOLD': 'strong',
  'ITALIC': 'em',
  'text': null,
  'unstyled': null,
};

function renderToken(token, index = 0, key = '0') {
  const { type, data, text, tokens } = token;
  const element = defaultRenderers[type];
  key = `${key}.${index}`;

  // Only render if type is registered as renderer
  if (typeof element !== 'undefined') {
    let children = null;
    if (Array.isArray(tokens) && tokens.length) {
      children = tokens.map((token, idx) => renderToken(token, idx, key));
    } else if (type === 'text') {
      children = text;
    }
    if (element !== null) {
      // If this is a react element
      return React.createElement(
        element,
        { key, ...data }, // Add key as a prop
        Array.isArray(children) && children.length === 1
          ? children[0] : children); // Pass single child if possible
    } else {
      // If this is a text node
      return children;
    }
  }
  return null;
}

export default class MarkitupReactRenderer extends React.Component {

  constructor(props) {
    super(props);
    const { syntax } = props;
    this.parser = new MarkupIt(syntax);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.syntax != this.props.syntax) {
      this.parser = new MarkupIt(nextProps.syntax);
    }
  }

  render() {
    const { value } = this.props;
    const content = this.parser.toContent(value);
    const json = JSONUtils.encode(content);
    // console.log(JSON.stringify(json, null, 2));

    return renderToken(json.token);
  }
}

MarkitupReactRenderer.propTypes = {
  value: PropTypes.string,
  syntax: PropTypes.instanceOf(Syntax).isRequired
};
