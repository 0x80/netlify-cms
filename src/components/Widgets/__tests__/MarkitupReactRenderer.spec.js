import React from 'react';
import { shallow } from 'enzyme';
import markdownSyntax from 'markup-it/syntaxes/markdown';
import htmlSyntax from 'markup-it/syntaxes/html';
import MarkitupReactRenderer from '../MarkitupReactRenderer';

describe('MarkitupReactRenderer', () => {
  it('should re-render properly after a value and syntax update', () => {
    const component = shallow(
      <MarkitupReactRenderer
          value="# Title"
          syntax={markdownSyntax}
      />
    );
    const tree1 = component.html();
    component.setProps({
      value: '<h1>Title</h1>',
      syntax: htmlSyntax
    });
    const tree2 = component.html();
    expect(tree1).toEqual(tree2);
  });

  it('should not update the parser if syntax didn\'t change', () => {
    const component = shallow(
      <MarkitupReactRenderer
          value="# Title"
          syntax={markdownSyntax}
      />
    );
    const syntax1 = component.instance().props.syntax;
    component.setProps({
      value: '## Title',
    });
    const syntax2 = component.instance().props.syntax;
    expect(syntax1).toEqual(syntax2);
  });

  it('should render markdown', () => {
    const value = `
# H1

Text with **bold** & _em_ elements

## H2

* ul item 1
* ul item 2

### H3

1. ol item 1
1. ol item 2
1. ol item 3

#### H4

[link title](http://google.com)

##### H5

![alt text](https://pbs.twimg.com/profile_images/678903331176214528/TQTdqGwD.jpg)

###### H6
`;
    const component = shallow(
      <MarkitupReactRenderer
          value={value}
          syntax={markdownSyntax}
      />
    );
    const tree = component.html();
    expect(tree).toMatchSnapshot();
  });

  it('should render HTML as is using Markdown', () => {
    const value = `
# Title

<dl>
    <dt>Test HTML content</dt>
    <dd>Testing HTML in Markdown</dd>
</dl>
`;
    const component = shallow(
      <MarkitupReactRenderer
          value={value}
          syntax={markdownSyntax}
      />
    );
    const tree = component.html();
    expect(tree).toMatchSnapshot();
  });

  it('should support custom syntax', () => {
    const value = '';
    const component = shallow(
      <MarkitupReactRenderer
          value={value}
          syntax={markdownSyntax}
      />
    );
    const tree = component.html();
    expect(tree).toMatchSnapshot();
  });

  it('should render HTML', () => {
    const value = '<p class="test class">Paragraph with <em>inline</em> element</p>';
    const component = shallow(
      <MarkitupReactRenderer
          value={value}
          syntax={htmlSyntax}
      />
    );
    const tree = component.html();
    expect(tree).toMatchSnapshot();
  });
});
