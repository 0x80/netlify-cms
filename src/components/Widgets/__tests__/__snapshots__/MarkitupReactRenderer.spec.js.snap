exports[`MarkitupReactRenderer should render HTML 1`] = `"<article><p>Paragraph with <em>inline</em> element</p></article>"`;

exports[`MarkitupReactRenderer should render HTML as is using Markdown 1`] = `"<article><h1>Title</h1></article>"`;

exports[`MarkitupReactRenderer should render markdown 1`] = `"<article><h1>H1</h1><p>Text with <strong>bold</strong> &amp; <em>em</em> elements</p><h2>H2</h2><ul><li>ul item 1</li><li>ul item 2</li></ul><h3>H3</h3><ol><li>ol item 1</li><li>ol item 2</li><li>ol item 3</li></ol><h4>H4</h4><p><a href=\"http://google.com\">link title</a></p><h5>H5</h5><p><img alt=\"alt text\" src=\"https://pbs.twimg.com/profile_images/678903331176214528/TQTdqGwD.jpg\"/></p><h6>H6</h6></article>"`;

exports[`MarkitupReactRenderer should support custom syntax 1`] = `"<article></article>"`;
